src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs

```diff
@@ -211,24 +209,90 @@ private static void EmitCallsToMetadataProvidersForParameters(this Endpoint endp
             }
 
-            if (parameter.IsEndpointParameterMetadataProvider)
+        }
+    }
+
+    public static void EmitFormAcceptsMetadata(this Endpoint endpoint, CodeWriter codeWriter)
+    {
+        var hasFormFiles = endpoint.Parameters.Any(p => p.IsFormFile);
+
+        if (hasFormFiles)
+        {
+            codeWriter.WriteLine("options.EndpointBuilder.Metadata.Add(new GeneratedAcceptsMetadata(contentTypes: GeneratedMetadataConstants.FormFileContentType));");
+        }
+        else
+        {
+            codeWriter.WriteLine("options.EndpointBuilder.Metadata.Add(new GeneratedAcceptsMetadata(contentTypes: GeneratedMetadataConstants.FormContentType));");
+        }
+    }
+
+    public static void EmitJsonAcceptsMetadata(this Endpoint endpoint, CodeWriter codeWriter)
+    {
+        codeWriter.WriteLine("var serviceProvider = options?.ServiceProvider ?? options?.EndpointBuilder?.ApplicationServices;");
+        codeWriter.WriteLine($"var serviceProviderIsService = serviceProvider?.GetService<IServiceProviderIsService>();");
+
+        if (endpoint.Parameters.SingleOrDefault(p => p.Source == EndpointParameterSource.JsonBody) is { } explicitBodyParameter)
+        {
+            // If we have a parameter which is explicitly set as the JSON body then we
+            // can assume its the one we want to build metadata from.
+            codeWriter.WriteLine($$"""options!.EndpointBuilder.Metadata.Add(new GeneratedAcceptsMetadata(type: typeof({{explicitBodyParameter.Type.ToDisplayString(EmitterConstants.DisplayFormatWithoutNullability)}}), isOptional: {{(explicitBodyParameter.IsOptional ? "true" : "false")}}, contentTypes: GeneratedMetadataConstants.JsonContentType));""");
+        }
+        else if (endpoint.Parameters.Any(p => p.Source == EndpointParameterSource.JsonBodyOrService))
+        {
+            var jsonBodyOrServiceParameters = endpoint.Parameters.Where(p => p.Source == EndpointParameterSource.JsonBodyOrService);
+
+            codeWriter.WriteLine("var jsonBodyOrServiceTypeTuples = new (bool, Type)[] {");
+            codeWriter.Indent++;
+            foreach (var parameter in jsonBodyOrServiceParameters)
             {
-                codeWriter.WriteLine($$"""var {{parameter.SymbolName}}_ParameterInfo = parameterInfos.Single(p => p.Name == "{{parameter.SymbolName}}");""");
-                codeWriter.WriteLine($"PopulateMetadataForParameter<{parameterType.ToDisplayString(EmitterConstants.DisplayFormat)}>({parameter.SymbolName}_ParameterInfo, options.EndpointBuilder);");
+                codeWriter.WriteLine($$"""({{(parameter.IsOptional ? "true" : "false")}}, typeof({{parameter.Type.ToDisplayString(EmitterConstants.DisplayFormatWithoutNullability)}})),""");
             }
+            codeWriter.Indent--;
+            codeWriter.WriteLine("};");
+            codeWriter.WriteLine($"var inferredBodyParameters = jsonBodyOrServiceTypeTuples.Where(p => !serviceProviderIsService!.IsService(p.Item2));");
```

Comment:
I wonder if we can reuse `EmitJsonBodyOrServiceResolver` here so that the resolution method that we emit can return two `Func`s. The existing one that is used to read from the request (`Func<HttpContext, bool, ValueTask<(bool, T?)>>`) and another once that knows how to populate the metadata object (`Action<IList<object>>`). That we way can reuse the `IServiceProviderIsService` check.

Reply:
I don't quite see the benefit of doing this. I think it conflates two chunks of code that are trying to do different things. Yes both chunks of code call `IsService(...)` but the execute at different parts of the lifecycle. I think it would make it more confusing.